File: quantra-backend/app/schemas/screener.py
"""
Pydantic v2 schemas — Screener.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class ScreenerFilter(BaseModel):
    field: str  # pe, rsi, ml_score, sentiment, direction, sector, ...
    op: str  # eq, neq, gt, gte, lt, lte, between, in
    value: Any  # single value, list, or [min, max] for between


class ScreenerRequest(BaseModel):
    filters: list[ScreenerFilter]
    sort_by: str = Field(default="ml_score")
    sort_order: str = Field(default="desc", pattern="^(asc|desc)$")
    limit: int = Field(default=30, ge=1, le=200)


class ScreenerResultItem(BaseModel):
    ticker: str
    name: str
    sector: str | None = None
    current_price: float | None = None
    pe_ratio: float | None = None
    rsi: float | None = None
    ml_score: float | None = None
    sentiment: float | None = None
    direction: str | None = None
    confidence: float | None = None


class ScreenerResponse(BaseModel):
    results: list[ScreenerResultItem]
    total: int
    filters_applied: int


class FilterFieldInfo(BaseModel):
    field: str
    label: str
    type: str  # numeric, text, enum
    min_value: float | None = None
    max_value: float | None = None
    options: list[str] | None = None


class PresetScreen(BaseModel):
    name: str
    description: str
    filters: list[ScreenerFilter]

File: quantra-frontend/src/types/screener.ts
export interface ScreenerFilter {
  field: string;
  operator: '>' | '<' | '>=' | '<=' | '=' | 'between' | 'in';
  value: string | number | number[] | string[];
}

export interface ScreenerRequest {
  filters: ScreenerFilter[];
  sort_by?: string;
  sort_dir?: 'asc' | 'desc';
  limit?: number;
}

export interface ScreenerResult {
  ticker: string;
  name: string;
  sector: string;
  price: number;
  change_pct: number;
  ml_score: number;
  direction: string;
  rsi: number | null;
  sentiment: number | null;
  pe: number | null;
  volume: number;
}

export interface ScreenerPreset {
  id: string;
  name: string;
  description: string;
  filters: ScreenerFilter[];
}

export const SCREENER_FIELDS = [
  { value: 'pe_ratio', label: 'P/E Ratio', type: 'number' },
  { value: 'pb_ratio', label: 'P/B Ratio', type: 'number' },
  { value: 'rsi_14', label: 'RSI (14)', type: 'number' },
  { value: 'macd_signal', label: 'MACD Signal', type: 'number' },
  { value: 'ml_score', label: 'ML Score', type: 'number' },
  { value: 'sentiment_score', label: 'Sentiment Score', type: 'number' },
  { value: 'market_cap', label: 'Market Cap', type: 'number' },
  { value: 'sector', label: 'Sector', type: 'string' },
  { value: 'direction', label: 'Direction', type: 'string' },
  { value: 'volume_spike', label: 'Volume Spike %', type: 'number' },
] as const;

export const SCREENER_OPERATORS = {
  number: ['>', '<', '>=', '<=', '=', 'between'],
  string: ['=', 'in'],
} as const;
